zcash_client_sqlite/src/wallet.rs

```diff
@@ -808,213 +808,449 @@ pub(crate) fn get_derived_account<P: consensus::Parameters>(
+                                ),
+                                BlockHeight::from_u32(row.get(1)?),
+                            ))
+                        },
+                    )
+                    // `None` if we have no subtree roots yet.
+                    .optional()?;
 
-            // We don't have complete information on how many outputs will exist in the shard at
-            // the chain tip without having scanned the chain tip block, so we overestimate by
-            // computing the maximum possible number of notes directly from the shard indices.
-            //
-            // TODO: it would be nice to be able to reliably have the size of the commitment tree
-            // at the chain tip without having to have scanned that block.
-            Ok(conn
-                .query_row(
-                    "SELECT MIN(shard_index), MAX(shard_index)
-                     FROM sapling_tree_shards
-                     WHERE subtree_end_height > :start_height
-                     OR subtree_end_height IS NULL",
-                    named_params![":start_height": u32::from(birthday_height)],
-                    |row| {
-                        let min_tree_size = row
-                            .get::<_, Option<u64>>(0)?
-                            .map(|min_idx| min_idx << SAPLING_SHARD_HEIGHT);
-                        let max_tree_size = row
-                            .get::<_, Option<u64>>(1)?
-                            .map(|max_idx| (max_idx + 1) << SAPLING_SHARD_HEIGHT);
-                        Ok(start_size.or(min_tree_size).zip(max_tree_size).map(
-                            |(min_tree_size, max_tree_size)| {
-                                Ratio::new(
-                                    scanned_count.unwrap_or(0),
-                                    max_tree_size - min_tree_size,
+                if let Some((last_completed_subtree, last_completed_subtree_end)) =
+                    last_completed_subtree
+                {
+                    // If we know the tree size at the last scanned height, and that
+                    // height is within the incomplete subtree, extrapolate.
+                    let tip_tree_size =
+                        last_scanned.and_then(|(last_scanned, last_scanned_tree_size)| {
+                            (last_scanned > last_completed_subtree_end)
+                                .then(|| {
+                                    let scanned_notes = last_scanned_tree_size
+                                        - u64::from(last_completed_subtree.position_range_end());
+                                    let scanned_range =
+                                        u64::from(last_scanned - last_completed_subtree_end);
+                                    let unscanned_range =
+                                        u64::from(chain_tip_height - last_scanned);
+
+                                    (scanned_notes * unscanned_range)
```

Comment:
saturating_mul? I think it can't overflow anyway, but that isn't particularly obvious.